cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -0,0 +1,164 @@
+import org.cryptimeleon.incentivesystem.cryptoprotocol.model.keys.provider.ProviderPublicKey;
+import org.cryptimeleon.math.expressions.exponent.ExponentConstantExpr;
+import org.cryptimeleon.math.expressions.exponent.ExponentMulExpr;
+import org.cryptimeleon.math.expressions.exponent.ExponentSumExpr;
+import org.cryptimeleon.math.expressions.group.GroupElementExpression;
+import org.cryptimeleon.math.structures.cartesian.GroupElementExpressionVector;
+import org.cryptimeleon.math.structures.cartesian.Vector;
+import org.cryptimeleon.math.structures.groups.Group;
+import org.cryptimeleon.math.structures.groups.GroupElement;
+import org.cryptimeleon.math.structures.rings.zn.Zn;
+
+import java.math.BigInteger;
+
+/**
+ * The ZKP used in the spend-deduct protocol
+ */
+public class SpendDeductZkp extends DelegateProtocol {
+
+    private final IncentivePublicParameters pp;
+    private final Zn zn;
+    private final Group groupG1;
+    private final ProviderPublicKey providerPublicKey;
+
+    public SpendDeductZkp(IncentivePublicParameters incentivePublicParameters, ProviderPublicKey providerPublicKey) {
+        this.providerPublicKey = providerPublicKey;
+        this.pp = incentivePublicParameters;
+        this.zn = incentivePublicParameters.getBg().getZn();
+        this.groupG1 = incentivePublicParameters.getBg().getG1();
+    }
+
+
+    @Override
+    protected SendThenDelegateFragment.SubprotocolSpec provideSubprotocolSpec(CommonInput pCommonInput, SendThenDelegateFragment.SubprotocolSpecBuilder builder) {
+        var commonInput = (SpendDeductZkpCommonInput) pCommonInput;
+        var H = new GroupElementExpressionVector(providerPublicKey.getH().pad(pp.getH7(), 7).map(GroupElement::expr));
+        var w = pp.getW();
+
+        // Variables to use
+        var eskVar = builder.addZnVariable("esk", zn);
+        var uskVar = builder.addZnVariable("usk", zn);
+        var dsrnd0Var = builder.addZnVariable("dsrnd0", zn);
+        var dsrnd1Var = builder.addZnVariable("dsrnd1", zn);
+        var dsrndStar0Var = builder.addZnVariable("dsrndStar0", zn);
+        var dsrndStar1Var = builder.addZnVariable("dsrndStar1", zn);
+        var vVar = builder.addZnVariable("v", zn);
+        var zVar = builder.addZnVariable("z", zn);
+        var tVar = builder.addZnVariable("t", zn);
+        var zStarVar = builder.addZnVariable("zStar", zn);
+        var tStarVar = builder.addZnVariable("tStar", zn);
+        var uStarVar = builder.addZnVariable("uStar", zn);
+        var uStarInverseVar = builder.addZnVariable("uStarInverse", zn);
+        var eskDecVarVector = new SchnorrZnVariable[pp.getNumEskDigits()];
+        var rVector = new SchnorrZnVariable[pp.getNumEskDigits()];
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            eskDecVarVector[i] = builder.addZnVariable("eskStarUserDec_" + i, zn);
+            rVector[i] = builder.addZnVariable("r_" + i, zn);
```

Comment:
This concatenation works without a cast? Looks like Java never stops surprising me...

Reply:
Which concatenation do you mean? String+int works

Reply:
As I sais, never stops surprising me :)